src/coreclr/jit/valuenum.cpp

```diff
@@ -8700,6 +8700,14 @@ void Compiler::fgValueNumberTree(GenTree* tree)
                         VNFunc loadFunc =
                             ((tree->gtFlags & GTF_IND_NONNULL) != 0) ? VNF_InvariantNonNullLoad : VNF_InvariantLoad;
 
+                        // Special case: for initialized non-null 'static readonly' fields we want to keep field
+                        // sequence to be able to fold their value
+                        if ((loadFunc == VNF_InvariantNonNullLoad) && addr->IsIconHandle(GTF_ICON_CONST_PTR) &&
+                            addr->AsIntCon()->gtFieldSeq != nullptr)
```

Comment:
```suggestion
                            (addr->AsIntCon()->gtFieldSeq != nullptr) &&
                            (addr->AsIntCon()->gtFieldSeq->GetOffset() == addr->AsIntCon()->IconValue())))
```
(This is what I meant by "zero offset" - it will ensure that we are only looking at the root field)